### ISSUE TITLE
`is_constant` does not work correctly for variable bounds with mutable params

### ISSUE THREAD

**Issue by michaelbynum** [Pyomo/pyomo#1744]
```
import pyomo.environ as pyo
m = pyo.ConcreteModel()
m.p = pyo.Param(mutable=True, initialize=5.2)
m.x = pyo.Var()
m.x.setlb(m.p)
assert not pyo.is_constant(m.x.lb) # fails
```

**Comment by ghackebeil** [comment]
I believe what you want is ‘not pyo.is_potentially_variable(...)’. Kernel might use one called is_data(...).

**Comment by ghackebeil** [comment]
And I believe at model write time, we relax this to only check is_fixed, which would potentially allow for a fixed variable to be used, but I don’t know why you would want to do that.

**Comment by michaelbynum** [comment]
`is_potentially_variable` returns False. I think the problem is that the `lb` property actually evaluates `_lb` to compare it to the bounds coming from the domain.

**Comment by michaelbynum** [comment]
The reason I care about this is for the persistent solver interfaces

**Comment by jsiirola** [comment]
@michaelbynum: that is correct (that `.lb` evaluates the lower bound to a numeric value), although I don't know why that is the case

**Comment by jsiirola** [comment]
Ahh - I get it.  The problem is that the actual lower bound of a variable could be coming from either the bound itself or the variable's domain.  `.lb` returns the tighter of the two.

**Comment by michaelbynum** [comment]
I think the problem is here:

https://github.com/Pyomo/pyomo/blob/b54f3991913c4a769196d3a509bb8e89c3e88841/pyomo/core/base/var.py#L402

You can't take the max if one of the arguments is not constant.

**Comment by michaelbynum** [comment]
Perhaps `setlb` and `setub` should check that their arguments are constant?

**Comment by jsiirola** [comment]
I don't think so - because it should be fine for any `not is_potentially_variable()` value to be a bound (people expect that you can use a mutable Param as a bound!).  But we should probably have an API that distinguishes between the effective lower bound "expression" (tighter of domain and bound) and the actual lower bound.  I fear implementing this for the 5.7.2 release, But I am definitely in favor of sorting it out before 6.0.

**Comment by michaelbynum** [comment]
That makes sense to me.

**Comment by michaelbynum** [comment]
We have discussed a few options to resolve this. I'll try to summarize and then make a proposal.

Current Behavior
-----------------
The bounds and domain are stored separately. When a user asks for the bounds (`v.lb`), we return the tighter of the bounds and the domain. In other words, both the bounds and the domain are enforced.

Alternative Options
--------------------
We could only allow unbounded domains. This would resolve this issue, but it is simply not possible for backwards compatibility reasons. 

Proposal
---------
1. Only allow constant values for variable bounds. In other words `v.setlb(m.p)` would cause an exception if `m.p` is mutable.
2. Raise exceptions if bounds specified are not compatible with the domain. For example, if a variable has a domain of `NonNegativeReals` and sets the lower bound to `None` or `-5`, an exception would be raised.
3. If the lower bound is `None` and the domain is `Reals`, changing the domain to `NonNegativeReals` would **not** raise an exception.

**Comment by ghackebeil** [comment]
My guess is that 1. would break many models.

**Comment by michaelbynum** [comment]
It is possible, but we currently get incorrect behavior. We did also discuss adding API to differentiate between the stored bounds and the implied bounds, but I think that is very confusing.

**Comment by ghackebeil** [comment]
I don't see the need to store both domain bounds and user bounds. The domain bounds are available immediately and can be transferred to the user bounds. The only case where this is problematic is when a user gives a bounded domain AND non-constant bounds (e.g., a mutable parameter that may are may not store a value yet). If think it's okay to throw an error in that case (may still break some models, but I think much fewer models). This a basically what kernel does.

**Comment by michaelbynum** [comment]
You are right. I like that idea.

**Comment by michaelbynum** [comment]
We discussed this. We think the best way to ensure correct behavior and maintain backwards compatibility is to add `min` and `max` expression nodes to the expression system. Then, if the domain is bounded, we can return an expression that depends on the domain. The `lb` and `ub` properties can return expressions if the bounds are not constant.

**Comment by michaelbynum** [comment]
This was resolved in #2172
